File: kafka-to-splunk-streaming-java/src/main/java/com/example/splunk/SplunkSender.java
package com.example.splunk;

import com.example.config.Config;
import org.json.JSONObject;

import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class SplunkSender {

    public static void sendToSplunk(String rawPayload) {
        try {
            JSONObject obj = new JSONObject(rawPayload);

String targetIndex;
if (obj.has("otel.splunkindex") && !obj.getString("otel.splunkindex").trim().isEmpty()) {
    String baseIndex = obj.getString("otel.splunkindex").trim();
    targetIndex = baseIndex + "_" + Config.CUSTOM_TEXT;

    // Optional: update the payload itself
    obj.put("otel.splunkindex", targetIndex);
} else {
    targetIndex = Config.DEFAULT_SPLUNK_INDEX;
}
            }

            JSONObject event = new JSONObject();
            event.put("event", obj);
            event.put("index", targetIndex);
            event.put("sourcetype", "_json");

            URL url = new URL(Config.SPLUNK_HEC_URL + "/services/collector/event");
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestMethod("POST");
            conn.setRequestProperty("Authorization", "Splunk " + Config.SPLUNK_TOKEN);
            conn.setRequestProperty("Content-Type", "application/json");
            conn.setDoOutput(true);

            try (OutputStream os = conn.getOutputStream()) {
                os.write(event.toString().getBytes());
                os.flush();
            }

            int code = conn.getResponseCode();
            if (code != 200) {
                System.err.println("⚠️ Splunk HEC failed: " + code + " - " + conn.getResponseMessage());
            }

        } catch (Exception e) {
            System.err.println("🚨 Error sending to Splunk: " + e.getMessage());
        }
    }
}

File: kafka_to_splunk_streaming/app/splunk_sender.py
import requests
import json
import logging
from app.config import SPLUNK_HEC_URL, SPLUNK_TOKEN, SPLUNK_DEFAULT_INDEX

logging.basicConfig(level=logging.INFO)

session = requests.Session()
headers = {
    'Authorization': f'Splunk {SPLUNK_TOKEN}',
    'Content-Type': 'application/json'
}

def send_to_splunk(payload_raw):
    try:
        data = json.loads(payload_raw)
    except json.JSONDecodeError:
        logging.warning("Invalid JSON")
        return

    target_index = f"{data.get('otel.splunkindex', '').strip()}_logs" if data.get("otel.splunkindex") else SPLUNK_DEFAULT_INDEX

    event = {
        "event": data,
        "index": target_index,
        "sourcetype": "_json"
    }

    try:
        r = session.post(f"{SPLUNK_HEC_URL}/services/collector/event", headers=headers, json=event, timeout=2)
        if r.status_code != 200:
            logging.warning(f"Splunk HEC failed: {r.status_code} - {r.text}")
    except Exception as e:
        logging.error(f"Splunk send error: {e}")
